BUG fish-lsp causing Helix editor to hang when saving changes and quitting

Apologies if this isn't a fish-lsp problem, it is hard to tell on which side the issue actually is, so if it isn't here then I'll open something on the Helix issues.

**Describe the bug**
When attempting to write and quit (`:wq`) Helix editor (as configured per the fish-lsp readme) it causes the editor to hang for about 20 seconds. It seems this is related to some formatting as if you instead use `:wq --no-format`, the editor quits instantly. Helix also generates logs indicating the same thing:

```log
2026-05-08T12:14:48.990 helix_view::document [WARN] LSP formatting failed: request 3 timed out
2026-05-08T12:14:58.292 helix_view::document [WARN] LSP formatting failed: request 6 timed out
2026-05-08T12:15:37.345 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(1)), original request likely timed out
2026-05-08T12:15:38.157 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(2)), original request likely timed out
2026-05-08T12:15:58.911 helix_view::document [WARN] LSP formatting failed: request 4 timed out
```

I did try running the fish-lsp logs as per the the troubleshooting section of the readme but it didn't seem to generate anything particularly useful.

**To Reproduce**
- Configure Helix as per fish-lsp readme
- Confirm installed etc. with `helix --health`
- Open a `.fish` file and add something basic like a function or just `echo 1` somewhere
- Save and quit with `:wq` (also happens if you separate them into `:w` and `:q`)
- Helix will hang for approx 20s and generate log file errors as shown above

You can mitigate this by quitting via `:wq --no-format` - no log entries are generated and Helix saves and quits instantly.

**Expected behavior**
Editor should close quickly and not generate errors.

**Screenshots**
N/A

**Please complete the following information:**
OS NAME: GNU/Linux (not overly useful, `lsb_release -d`: Fedora release 43)
YARN VERSION: 1.22.19
NODE VERSION: v22.2.0
FISH VERSION: fish, version 4.7.1
FISH-LSP VERSION: 1.1.3

**Additional context**
Using Helix editor: `helix 25.07.1 (7d92fbda)` (from very recent `master` branch commits)
I did also try to see if I could reproduce this on other CLI editors - namely Nvim or Kakoune but having little experience with those I don't think I ever got either set up correctly due to inexperience with the plugin system for both.

I do have some slightly less easily reproducible info. Sometimes if splitting `:w` and `:q` it will cause Helix to fail to quit at all with a message about the file being modified without saving. I have also seen some other messages pop up in the Helix log but cannot get them to pop up all the time:

```log
2026-05-08T12:23:13.386 helix_lsp::client [WARN] language server failed to terminate gracefully - server closed the stream
2026-05-08T12:23:13.386 helix_lsp::client [ERROR] Failed to send notification 'exit' to server 'fish-lsp': channel closed 
```

fish-lsp generated log below, nothing really useful showed up that I could find.
<details><summary>log file</summary>
<pre><code>
{
  "onInitializedResult": {
    "capabilities": {
      "textDocumentSync": {
        "openClose": true,
        "change": 2,
        "save": {
          "includeText": true
        }
      },
      "completionProvider": {
        "resolveProvider": true,
        "allCommitCharacters": [
          "\t",
          ";",
          " "
        ],
        "workDoneProgress": false
      },
      "hoverProvider": true,
      "definitionProvider": true,
      "implementationProvider": true,
      "referencesProvider": true,
      "renameProvider": true,
      "documentFormattingProvider": true,
      "documentRangeFormattingProvider": true,
      "foldingRangeProvider": true,
      "selectionRangeProvider": true,
      "codeActionProvider": {
        "codeActionKinds": [
          "quickfix.fix",
          "quickfix.disable",
          "quickfix.fixAll",
          "quickfix.delete",
          "refactor.rewrite",
          "refactor.extract",
          "source.rename"
        ],
        "workDoneProgress": true,
        "resolveProvider": true
      },
      "executeCommandProvider": {
        "commands": [
          "quickfix.fix",
          "quickfix.disable",
          "quickfix.fixAll",
          "quickfix.delete",
          "refactor.rewrite",
          "refactor.extract",
          "source.rename",
          "fish-lsp.executeRange",
          "fish-lsp.executeLine",
          "fish-lsp.execute",
          "fish-lsp.executeBuffer",
          "fish-lsp.createTheme",
          "fish-lsp.showStatusDocs",
          "fish-lsp.showWorkspaceMessage",
          "fish-lsp.updateWorkspace",
          "fish-lsp.fixAll",
          "fish-lsp.toggleSingleWorkspaceSupport",
          "fish-lsp.generateEnvVariables",
          "fish-lsp.showEnvVariables",
          "fish-lsp.checkHealth",
          "fish-lsp.showReferences",
          "fish-lsp.showInfo"
        ],
        "workDoneProgress": true
      },
      "documentSymbolProvider": {
        "label": "fish-lsp"
      },
      "workspaceSymbolProvider": {
        "resolveProvider": true
      },
      "documentHighlightProvider": true,
      "inlayHintProvider": true,
      "semanticTokensProvider": {
        "legend": {
          "tokenTypes": [
            "function",
            "variable",
            "keyword",
            "operator",
            "decorator",
            "string",
            "number",
            "event"
          ],
          "tokenModifiers": [
            "local",
            "inherit",
            "function",
            "global",
            "universal",
            "export",
            "defaultLibrary",
            "builtin"
          ]
        },
        "range": true,
        "full": {
          "delta": false
        }
      },
      "signatureHelpProvider": {
        "workDoneProgress": false,
        "triggerCharacters": [
          "."
        ]
      },
      "documentOnTypeFormattingProvider": {
        "firstTriggerCharacter": ".",
        "moreTriggerCharacter": [
          ";",
          "}",
          "]",
          ")"
        ]
      },
      "workspace": {
        "workspaceFolders": {
          "supported": true,
          "changeNotifications": true
        }
      }
    },
    "serverInfo": {
      "name": "fish-lsp",
      "version": "1.1.3",
      "buildTime": "12/18/25, 1:57:40 AM",
      "buildPath": "/home/dae/.config/fish/functions/fish-lsp"
    }
  }
}
{
  "server": "FishServer",
  "rootUri": null,
  "rootPath": "/home/dae/.config/fish/functions",
  "workspaceFolders": []
}
DEBUG: hasWorkspaceFolderCapability
true
getWorkspacePathsFromInitializationParams(params)
{
  "rootUri": null,
  "rootPath": "/home/dae/.config/fish/functions",
  "workspaceFolders": []
}
INFO: initializeUris
["/home/dae/.config/fish/functions"]
DEBUG: getWorkspaceName
{
  "root": "/home/dae/.config/fish",
  "specialName": "__fish_config_dir"
}
DEBUG: getWorkspaceName
{
  "root": "/home/dae/.config/fish",
  "specialName": "__fish_config_dir"
}
INFO: Initialized workspace '__fish_config_dir' @ 0
{
  "name": "__fish_config_dir",
  "uri": "file:///home/dae/.config/fish",
  "path": "/home/dae/.config/fish"
}
</code></pre>
</details>

My Helix setup is pretty vanilla and really not been customised much. I tried that change to the config as well as setting the global variable but neither made a difference. Still getting the same kind of output.

What is of interest is that I tried a totally vanilla setup on another computer with only a new version of helix and download of fish-lsp and see something very similar but greatly reduced. On the other computer it fails in a similar way but only on the order of 5 seconds or so then prints the following to the shell.

> Error: Timed out waiting for language servers to shutdown

Apologies if this is purely a Helix issue and I'll raise it there if it is, just hard to tell which side the issue is on.

I should be able to have a proper dig through the logs tomorrow if it is still something of interest here.

The other logs do show some slight differences though:
```
2026-05-09T02:28:11.808 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(1)), original request likely timed out
2026-05-09T02:28:12.719 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(2)), original request likely timed out
2026-05-09T02:28:12.858 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(3)), original request likely timed out
2026-05-09T02:28:12.923 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(4)), original request likely timed out
2026-05-09T02:28:13.118 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(6)), original request likely timed out
2026-05-09T02:28:13.163 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(7)), original request likely timed out
2026-05-09T02:28:13.233 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(8)), original request likely timed out
2026-05-09T02:28:13.301 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(9)), original request likely timed out
2026-05-09T02:28:13.395 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(10)), original request likely timed out
2026-05-09T02:28:13.659 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(5)), original request likely timed out
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "Error: write EPIPE\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at afterWriteDispatched (node:internal/stream_base_commons:160:15)\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at writeGeneric (node:internal/stream_base_commons:151:3)\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at Socket._writeGeneric (node:net:923:11)\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at Socket._write (node:net:935:8)\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at writeOrBuffer (node:internal/streams/writable:392:12)\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at _write (node:internal/streams/writable:333:10)\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at Socket.Writable.write (node:internal/streams/writable:337:10)\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at /home/cpi/.local/src/formatting.ts:17:21\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at new Promise (<anonymous>)\n"
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at cl (/home/cpi/.local/src/formatting.ts:7:10)\n"
2026-05-09T02:28:16.270 helix_lsp::transport [ERROR] fish-lsp err: <- StreamClosed
2026-05-09T02:28:16.270 helix_view::document [WARN] LSP formatting failed: server closed the stream
2026-05-09T02:28:16.297 helix_lsp::transport [ERROR] fish-lsp err: <- IO(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
2026-05-09T02:28:16.297 helix_lsp::transport [ERROR] fish-lsp err: <- IO(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
2026-05-09T02:28:19.299 helix_term::application [ERROR] Timed out waiting for language servers to shutdown
```

Ahh okay, I think I those logs include where the error is coming from, but I still can't reproduce the issue you're having on my machine so I'm uncertain if any fixes I write entirely correct the problem.

In the logs you sent the stack trace mentions the error is being thrown in `src/formatting.ts#17` in message:

```
2026-05-09T02:28:16.101 helix_lsp::transport [ERROR] fish-lsp err <- "    at /home/cpi/.local/src/formatting.ts:17:21\n"
```

The fix needs to prevent `EPIPE` from escaping/crashing and fix the old behavior where formatter errors logged w/o leaving a promise hanging  

Would you mind testing if a build from branch [fix/issue-159](https://github.com/ndonfris/fish-lsp/tree/fix%2Fissue-159) corrects this issue?

--- a/scripts/esbuild/configs.ts
+++ b/scripts/esbuild/configs.ts
@@ -89,7 +89,6 @@ export const buildConfigs: Record<BuildConfigTarget, BuildConfig> = {
     // External dependencies - don't bundle these, npm will provide them
     external: [
       '@esdmr/tree-sitter-fish',
-      'chalk',
       'commander',
       'fast-glob',
       'fs-extra',

--- a/src/formatting.ts
+++ b/src/formatting.ts
@@ -1,40 +1,81 @@
-import { exec } from 'child_process';
+import { execFile } from 'child_process';
 import { logger } from './logger';
 import { LspDocument } from './document';
 import { getEnabledIndentRanges } from './parsing/comments';
 
-export async function formatDocumentContent(content: string): Promise<string> {
-  return new Promise((resolve, _reject) => {
-    const process = exec('fish_indent', (error, stdout, stderr) => {
-      if (error) {
-        // reject(stderr);
-        logger.log('Formatting Error:', stderr);
-      } else {
-        resolve(stdout);
-      }
-    });
-    if (process.stdin) {
-      process.stdin.write(content);
-      process.stdin.end();
+// Cap fish_indent at a generous wall-clock so a hung child cannot hold the LSP
+// formatting request open past a client's request timeout (e.g. helix' ~20s).
+const FISH_INDENT_TIMEOUT_MS = 10_000;
+// fish_indent output is bounded by input size, but Node's default execFile
+// maxBuffer is 1 MiB which truncates large fish files. 50 MiB is well clear of
+// any realistic source while still capping pathological inputs.
+const FISH_INDENT_MAX_BUFFER = 50 * 1024 * 1024;
+
+function runFishIndent(content: string, args: string[] = []): Promise<string> {
+  return new Promise((resolve, reject) => {
+    let settled = false;
+
+    const resolveOnce = (value: string) => {
+      if (settled) return;
+      settled = true;
+      resolve(value);
+    };
+
+    const rejectOnce = (error: Error) => {
+      if (settled) return;
+      settled = true;
+      logger.log('Formatting Error:', error.message);
+      reject(error);
+    };
+
+    const process = execFile(
+      'fish_indent',
+      args,
+      { timeout: FISH_INDENT_TIMEOUT_MS, maxBuffer: FISH_INDENT_MAX_BUFFER },
+      (error, stdout, stderr) => {
+        if (error) {
+          const err = error as NodeJS.ErrnoException & { killed?: boolean; signal?: string; };
+          let message: string;
+          if (err.killed && err.signal === 'SIGTERM') {
+            message = `fish_indent timed out after ${FISH_INDENT_TIMEOUT_MS}ms`;
+          } else if (err.code === 'ERR_CHILD_PROCESS_STDIO_MAXBUFFER') {
+            message = `fish_indent output exceeded ${FISH_INDENT_MAX_BUFFER} bytes`;
+          } else {
+            message = stderr.trim() || err.message;
+          }
+          rejectOnce(new Error(message));
+          return;
+        }
+
+        resolveOnce(stdout);
+      },
+    );
+
+    process.on('error', rejectOnce);
+
+    if (!process.stdin) {
+      rejectOnce(new Error('Unable to write to fish_indent stdin'));
+      process.kill();
+      return;
+    }
+
+    process.stdin.on('error', rejectOnce);
+
+    try {
+      process.stdin.end(content);
+    } catch (error) {
+      rejectOnce(error instanceof Error ? error : new Error(String(error)));
+      process.kill();
     }
   });
 }
 
+export async function formatDocumentContent(content: string): Promise<string> {
+  return runFishIndent(content);
+}
+
 export async function formatDocumentRangeContent(content: string): Promise<string> {
-  return new Promise((resolve, _reject) => {
-    const process = exec('fish_indent --only-indent --only-unindent', (error, stdout, stderr) => {
-      if (error) {
-        // reject(stderr);
-        logger.log('Formatting Error:', stderr);
-      } else {
-        resolve(stdout);
-      }
-    });
-    if (process.stdin) {
-      process.stdin.write(content);
-      process.stdin.end();
-    }
-  });
+  return runFishIndent(content, ['--only-indent', '--only-unindent']);
 }
 
 interface OriginalRange {

--- a/tests/formatting-child-process.test.ts
+++ b/tests/formatting-child-process.test.ts
@@ -1,0 +1,184 @@
+import { EventEmitter } from 'events';
+import { Writable } from 'stream';
+
+type ExecFileCallback = (error: Error | null, stdout: string, stderr: string) => void;
+type ExecFileOptions = { timeout?: number; maxBuffer?: number; };
+type ExecFileMock = (cmd: string, args: string[], options: ExecFileOptions, callback: ExecFileCallback) => unknown;
+
+function createMockProcess(stdin: Writable | null) {
+  const childProcess = new EventEmitter() as EventEmitter & {
+    stdin: Writable | null;
+    kill: ReturnType<typeof vi.fn>;
+  };
+
+  childProcess.stdin = stdin;
+  childProcess.kill = vi.fn();
+
+  return childProcess;
+}
+
+function mockFormattingDependencies() {
+  vi.doMock('../src/logger', () => ({
+    logger: {
+      log: vi.fn(),
+    },
+  }));
+  vi.doMock('../src/parsing/comments', () => ({
+    getEnabledIndentRanges: vi.fn(),
+  }));
+}
+
+describe('formatting child process failures', () => {
+  afterEach(() => {
+    vi.doUnmock('child_process');
+    vi.doUnmock('../src/logger');
+    vi.doUnmock('../src/parsing/comments');
+    vi.resetModules();
+  });
+
+  it('rejects when fish_indent closes stdin before content is written', async () => {
+    vi.resetModules();
+    mockFormattingDependencies();
+
+    const execFile: ExecFileMock = vi.fn((_cmd, _args, _options, _callback) => {
+      const stdin = new Writable({
+        write(_chunk, _encoding, callback) {
+          const error = new Error('write EPIPE') as NodeJS.ErrnoException;
+          error.code = 'EPIPE';
+          callback(error);
+        },
+      });
+
+      return createMockProcess(stdin);
+    });
+
+    vi.doMock('child_process', async (importOriginal) => {
+      const actual = await importOriginal<typeof import('child_process')>();
+      return { ...actual, execFile };
+    });
+
+    const { formatDocumentContent } = await import('../src/formatting');
+
+    await expect(formatDocumentContent('echo test')).rejects.toThrow('write EPIPE');
+  });
+
+  it('rejects when fish_indent exits with an error', async () => {
+    vi.resetModules();
+    mockFormattingDependencies();
+
+    const execFile: ExecFileMock = vi.fn((_cmd, _args, _options, callback) => {
+      const stdin = new Writable({
+        write(_chunk, _encoding, writeComplete) {
+          writeComplete();
+        },
+      });
+      const childProcess = createMockProcess(stdin);
+
+      setImmediate(() => callback(new Error('fish_indent failed'), '', 'invalid fish syntax'));
+
+      return childProcess;
+    });
+
+    vi.doMock('child_process', async (importOriginal) => {
+      const actual = await importOriginal<typeof import('child_process')>();
+      return { ...actual, execFile };
+    });
+
+    const { formatDocumentContent } = await import('../src/formatting');
+
+    await expect(formatDocumentContent('if')).rejects.toThrow('invalid fish syntax');
+  });
+
+  it('passes a timeout and maxBuffer to execFile so a hung child cannot stall the LSP', async () => {
+    vi.resetModules();
+    mockFormattingDependencies();
+
+    const seenOptions: ExecFileOptions[] = [];
+    const execFile: ExecFileMock = vi.fn((_cmd, _args, options, callback) => {
+      seenOptions.push(options);
+      const stdin = new Writable({
+        write(_chunk, _encoding, writeComplete) {
+          writeComplete();
+        },
+      });
+      const childProcess = createMockProcess(stdin);
+      setImmediate(() => callback(null, 'echo ok\n', ''));
+      return childProcess;
+    });
+
+    vi.doMock('child_process', async (importOriginal) => {
+      const actual = await importOriginal<typeof import('child_process')>();
+      return { ...actual, execFile };
+    });
+
+    const { formatDocumentContent } = await import('../src/formatting');
+    await formatDocumentContent('echo ok');
+
+    expect(seenOptions).toHaveLength(1);
+    expect(seenOptions[0]?.timeout).toBeGreaterThan(0);
+    expect(seenOptions[0]?.maxBuffer).toBeGreaterThan(1024 * 1024);
+  });
+
+  it('reports a clear timeout error when execFile kills fish_indent for exceeding the deadline', async () => {
+    vi.resetModules();
+    mockFormattingDependencies();
+
+    const execFile: ExecFileMock = vi.fn((_cmd, _args, _options, callback) => {
+      const stdin = new Writable({
+        write(_chunk, _encoding, writeComplete) {
+          writeComplete();
+        },
+      });
+      const childProcess = createMockProcess(stdin);
+
+      setImmediate(() => {
+        const err = new Error('Command failed') as NodeJS.ErrnoException & { killed: boolean; signal: NodeJS.Signals; };
+        err.killed = true;
+        err.signal = 'SIGTERM';
+        callback(err, '', '');
+      });
+
+      return childProcess;
+    });
+
+    vi.doMock('child_process', async (importOriginal) => {
+      const actual = await importOriginal<typeof import('child_process')>();
+      return { ...actual, execFile };
+    });
+
+    const { formatDocumentContent } = await import('../src/formatting');
+
+    await expect(formatDocumentContent('echo waiting')).rejects.toThrow(/timed out/i);
+  });
+
+  it('reports a clear maxBuffer error when fish_indent output exceeds the buffer cap', async () => {
+    vi.resetModules();
+    mockFormattingDependencies();
+
+    const execFile: ExecFileMock = vi.fn((_cmd, _args, _options, callback) => {
+      const stdin = new Writable({
+        write(_chunk, _encoding, writeComplete) {
+          writeComplete();
+        },
+      });
+      const childProcess = createMockProcess(stdin);
+
+      setImmediate(() => {
+        const err = new Error('stdout maxBuffer length exceeded') as NodeJS.ErrnoException;
+        err.code = 'ERR_CHILD_PROCESS_STDIO_MAXBUFFER';
+        callback(err, '', '');
+      });
+
+      return childProcess;
+    });
+
+    vi.doMock('child_process', async (importOriginal) => {
+      const actual = await importOriginal<typeof import('child_process')>();
+      return { ...actual, execFile };
+    });
+
+    const { formatDocumentContent } = await import('../src/formatting');
+
+    await expect(formatDocumentContent('echo too-much')).rejects.toThrow(/exceeded/i);
+  });
+});

